crates/interpreter/src/frame.rs

```diff
@@ -0,0 +1,176 @@
+use std::mem;
+
+use cranelift_entity::SecondaryMap;
+
+use sonatina_ir::{
+    module::ModuleCtx,
+    types::{CompoundType, CompoundTypeData},
+    DataFlowGraph, Function, Insn, Type, Value, ValueData, I256,
+};
+
+use crate::{Literal, ProgramCounter};
+
+pub struct Frame {
+    pub ret_addr: ProgramCounter,
+    local_values: SecondaryMap<Value, Literal>, // 256-bit register
+    alloca_region: Vec<u8>,                     // big endian
+}
+
+impl Frame {
+    pub fn new(func: &Function, ret_addr: ProgramCounter, args: Vec<Literal>) -> Self {
+        let mut local_values = SecondaryMap::new();
+        for (v, literal_value) in func.arg_values.iter().zip(args.into_iter()) {
+            local_values[*v] = literal_value
+        }
+        let alloca_region = Vec::new();
+
+        Self {
+            ret_addr,
+            local_values,
+            alloca_region,
+        }
+    }
+
+    pub fn load(&mut self, /*ctx: Context,*/ v: Value, dfg: &DataFlowGraph) -> Literal {
+        if !self.is_assigned(v) {
+            let v = match dfg.value_data(v) {
+                ValueData::Insn { insn, .. } => {
+                    let result_v = dfg.insn_result(*insn).unwrap();
+                    if self.is_assigned(result_v) {
+                        return self.local_values[result_v];
+                    }
+                    result_v
+                }
+                _ => v,
+            };
+            let i256 = dfg.value_imm(v).unwrap().as_i256();
+            self.local_values[v] = Literal(i256);
+        }
```

Comment:
```suggestion
    let v = dfg.resolve_alias(v);
    if let Some(imm) = dfg.value_imm(v) {
        return imm.as_i256();
    }
```
1. `result_v` is always the same as `v`. So the `ValueData::Insn` arm is redundant.
2.  when `v` is `ValueData::Global`, this would cause a bug (We could handle global variables properly when we change the memory layout model).
3. We need to resolve an alias value before loading the value.

Reply:
ok, thank you. in number 3, `value_imm` calls `resolve_alias`.